Sources/NIO/Codec.swift

```diff
@@ -101,115 +129,266 @@ private extension ChannelHandlerContext {
-            ctx.fireChannelRead(self.wrapInboundOut(buffer))
-        } else {
-            /* please note that we're dropping the partially received bytes (if any) on the floor here as we can't
-               send a full message to the next handler. */
+    internal private(set) var decoder: Decoder?
+    private var state: State = .active
+    private var buffer: B2MDBuffer = B2MDBuffer()
+
+    public init(_ decoder: Decoder) {
+        self.decoder = decoder
+    }
+
+    deinit {
+        assert(self.state == .done)
+    }
+}
+
+// MARK: ByteToMessageHandler: Test Helpers
+extension ByteToMessageHandler {
+    internal var cumulationBuffer: ByteBuffer? {
+        return self.buffer._testOnlyOneBuffer()
+    }
+}
+
+// MARK: ByteToMessageHandler's Main API
+extension ByteToMessageHandler {
+    private func withNextBuffer(extraData: ByteBuffer?, _ body: (inout Decoder, inout ByteBuffer) throws -> DecodingState) rethrows -> B2MDBuffer.BufferProcessingResult {
+        if let extraData = extraData {
+            self.buffer.append(buffer: extraData)
+        }
+        switch self.buffer.startProcessing() {
+        case .bufferAlreadyBeingProcessed:
+            return .cannotProcessReentrantly
+        case .nothingAvailable:
+            return .didProcess(.needMoreData)
+        case .available(var buffer):
+            var decoder: Decoder? = nil
+            swap(&decoder, &self.decoder)
+            defer {
+                swap(&decoder, &self.decoder)
+                if buffer.readableBytes > 0 {
+                    if self.decoder!.shouldReclaimBytes(buffer: buffer) {
+                        buffer.discardReadBytes()
+                    }
+                }
+                self.buffer.finishProcessing(remainder: buffer)
+            }
+            return .didProcess(try body(&decoder!, &buffer))
         }
-        self.cumulationBuffer = nil
-        self.decoderRemoved(ctx: ctx)
     }
 
-    /// Just call `decode`. Users may implement their own logic.
-    public func decodeLast(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
-        return try self.decode(ctx: ctx, buffer: &buffer)
+    private func processLeftovers(ctx: ChannelHandlerContext) {
+        guard self.state == .active else {
+            // we are processing or have already processed the leftovers
+            return
+        }
+
+        ctx.withThrowingToFireErrorAndClose {
+            switch try self.decodeLoop(ctx: ctx, decodeMode: .last) {
+            case .didProcess:
+                self.state = .done
+            case .cannotProcessReentrantly:
+                self.state = .leftoversNeedProcessing
+            }
+        }
     }
 
-    /// Do nothing by default.
-    public func decoderRemoved(ctx: ChannelHandlerContext) {
+    private func decodeLoop(ctx: ChannelHandlerContext, extraData: ByteBuffer? = nil, decodeMode: DecodeMode) throws -> B2MDBuffer.BufferProcessingResult {
```

Comment:
I don't think this `extraData` parameter makes anything particularly clearer. Any reason not to just `append` it to `self.buffer` in `channelRead`?

Reply:
that's a good idea, thanks!